pkg/restore/log_client.go

```diff
@@ -0,0 +1,1079 @@
+
+type grpcClis struct {
+	mu   sync.Mutex
+	clis map[uint64]*grpc.ClientConn
+}
+
+// LogClient sends requests to restore files.
+type LogClient struct {
+	// lock DDL execution
+	// TODO remove lock by using db session pool if necessary
+	ddlLock sync.Mutex
+
+	restoreClient *Client
+	splitClient   SplitClient
+
+	// range of log backup
+	startTS uint64
+	endTs   uint64
+
+	concurrencyCfg concurrencyCfg
+	// meta info parsed from log backup
+	meta         *LogMeta
+	eventPullers map[int64]*cdclog.EventPuller
+	tableBuffers map[int64]*cdclog.TableBuffer
+	grpcClis     grpcClis
+
+	tableFilter filter.Filter
+
+	// a map to store all drop schema ts, use it as a filter
+	dropTSMap sync.Map
+}
+
+// NewLogRestoreClient returns a new LogRestoreClient.
+func NewLogRestoreClient(
+	ctx context.Context,
+	restoreClient *Client,
+	startTs uint64,
+	endTS uint64,
+	tableFilter filter.Filter,
+	concurrency uint,
+	batchFlushPairs int,
+	batchFlushSize int64,
+	batchWriteKVPairs int,
+) (*LogClient, error) {
+	var err error
+	if endTS == 0 {
+		// means restore all log data,
+		// so we get current ts from restore cluster
+		endTS, err = restoreClient.GetTS(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	splitClient := NewSplitClient(restoreClient.GetPDClient(), restoreClient.GetTLSConfig())
+
+	cfg := concurrencyCfg{
+		Concurrency:       concurrency,
+		BatchFlushKVPairs: batchFlushPairs,
+		BatchFlushKVSize:  batchFlushSize,
+		BatchWriteKVPairs: batchWriteKVPairs,
+	}
+
+	lc := &LogClient{
+		restoreClient:  restoreClient,
+		splitClient:    splitClient,
+		startTS:        startTs,
+		endTs:          endTS,
+		concurrencyCfg: cfg,
+		meta:           new(LogMeta),
+		eventPullers:   make(map[int64]*cdclog.EventPuller),
+		tableBuffers:   make(map[int64]*cdclog.TableBuffer),
+		tableFilter:    tableFilter,
+	}
+	lc.grpcClis.clis = make(map[uint64]*grpc.ClientConn)
+	return lc, nil
+}
+
+func (l *LogClient) getGrpcConnLocked(ctx context.Context, storeID uint64) (*grpc.ClientConn, error) {
+	store, err := l.splitClient.GetStore(ctx, storeID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	opt := grpc.WithInsecure()
+	if l.restoreClient.tlsConf != nil {
+		opt = grpc.WithTransportCredentials(credentials.NewTLS(l.restoreClient.tlsConf))
+	}
+	ctx, cancel := context.WithTimeout(ctx, dialTimeout)
+
+	bfConf := backoff.DefaultConfig
+	bfConf.MaxDelay = gRPCBackOffMaxDelay
+	conn, err := grpc.DialContext(
+		ctx,
+		store.GetAddress(),
+		opt,
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: bfConf}),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                gRPCKeepAliveTime,
+			Timeout:             gRPCKeepAliveTimeout,
+			PermitWithoutStream: true,
+		}),
+	)
+	cancel()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	// Cache the conn.
+	l.grpcClis.clis[storeID] = conn
+	return conn, nil
+}
+
+func (l *LogClient) getImportClient(ctx context.Context, peer *metapb.Peer) (sst.ImportSSTClient, error) {
```

Comment:
why are we not reusing `pkg/restore/import.go`

Reply:
It can be reused, I moved the grpc config setting in `pkg/restore/import.go` in this commit https://github.com/pingcap/br/pull/482/commits/1b6de45d4d111186e3eb43ee786816658e34d52d